engine/crates/engine-v2/src/execution/planner/mod.rs

```diff
@@ -1,131 +1,217 @@
+                let child_id = self.logical_plan_to_execution_plan_id[usize::from(child_logical_plan_id)]
+                    .expect("Depend on unfinished plan");
+                if !self.operation[parent_id].children.contains(&child_id) {
+                    self.operation[parent_id].children.push(child_id);
+                    self.operation[child_id].parent_count += 1;
+                }
+            }
+        }
 
-        for (i, dependencies) in execution_plans_dependencies.into_iter().enumerate() {
+        // child/parent relations between execution plans.
+        for (i, input_fields) in take(&mut self.execution_plans_input_fields).into_iter().enumerate() {
             let child_id = ExecutionPlanId::from(i);
-            for field_id in dependencies {
-                let logical_plan_id = operation.plan.field_to_logical_plan_id[usize::from(field_id)];
-                let parent_id =
-                    logical_plan_to_execution_plan_id[usize::from(logical_plan_id)].expect("Depend on unfinished plan");
-                if !operation[parent_id].children.contains(&child_id) {
-                    operation[parent_id].children.push(child_id);
-                    operation[child_id].parent_count += 1;
+            for &field_id in &self.build_context[input_fields] {
+                let parent_logical_plan_id = self.operation.plan.field_to_logical_plan_id[usize::from(field_id)];
+                let parent_id = self.logical_plan_to_execution_plan_id[usize::from(parent_logical_plan_id)]
+                    .expect("Depend on unfinished plan");
+                if !self.operation[parent_id].children.contains(&child_id) {
+                    self.operation[parent_id].children.push(child_id);
+                    self.operation[child_id].parent_count += 1;
                 }
             }
         }
 
         // To enforce the proper ordering of mutation fields, we create parent/child relations
         // between them.
-        let mut mutation_fields_plan_order = operation
+        let mut mutation_fields_plan_order = self
+            .operation
             .plan
             .mutation_fields_plan_order
             .clone()
             .into_iter()
-            .filter_map(|id| logical_plan_to_execution_plan_id[usize::from(id)]);
+            .filter_map(|id| self.logical_plan_to_execution_plan_id[usize::from(id)])
+            .collect::<Vec<_>>()
+            .into_iter();
         if let Some(mut prev) = mutation_fields_plan_order.next() {
             for next in mutation_fields_plan_order {
-                if !operation[prev].children.contains(&next) {
-                    operation[prev].children.push(next);
-                    operation[next].parent_count += 1;
+                if !self.operation[prev].children.contains(&next) {
+                    self.operation[prev].children.push(next);
+                    self.operation[next].parent_count += 1;
                 }
                 prev = next;
```

Comment:
This kind of looks like [tuple_windows](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.tuple_windows), might be easier to just use that (if i'm reading right)